File: Client/Gamify.Client.Net/Gamify.Client.Net/Managers/GameMoveManager.cs
using Gamify.Client.Net.Client;
using Gamify.Client.Net.Services;
using Gamify.Contracts.Notifications;
using Gamify.Contracts.Requests;
using System;

namespace Gamify.Client.Net.Managers
{
    public class GameMoveManager<TGameMoveRequest, TMoveResultNotification, TMoveNotification>
        where TGameMoveRequest : IRequestObject
        where TMoveResultNotification : INotificationObject
        where TMoveNotification : INotificationObject
    {
        private readonly IGameService<TGameMoveRequest, TMoveResultNotification> gameMoveService;
        private readonly IGameListener<TMoveNotification> gameMoveListener;

        public event EventHandler<GameNotificationEventArgs<TMoveResultNotification>> GameMoveResultNotificationReceived;
        public event EventHandler<GameNotificationEventArgs<TMoveNotification>> GameMoveNotificationReceived;

        public GameMoveManager(string playerName, IGameClientFactory clientFactory)
        {
            var gameClient = clientFactory.GetGameClient(playerName);

            this.gameMoveService = new GameService<TGameMoveRequest, TMoveResultNotification>(GameRequestType.GameMove, GameNotificationType.GameMoveResult, gameClient);
            this.gameMoveListener = new GameListener<TMoveNotification>(GameNotificationType.GameMove, gameClient);

            this.gameMoveService.NotificationReceived += (sender, args) =>
            {
                this.NotifyGameMoveResult(args);
            };
            this.gameMoveListener.NotificationReceived += (sender, args) =>
            {
                this.NotifyGameMove(args);
            };
        }

        public void SendMove(TGameMoveRequest gameMoveRequest)
        {
            this.gameMoveService.Send(gameMoveRequest);
        }

        private void NotifyGameMoveResult(GameNotificationEventArgs<TMoveResultNotification> args)
        {
            if (this.GameMoveResultNotificationReceived != null)
            {
                this.GameMoveResultNotificationReceived(this, args);
            }
        }

        private void NotifyGameMove(GameNotificationEventArgs<TMoveNotification> args)
        {
            if (this.GameMoveNotificationReceived != null)
            {
                this.GameMoveNotificationReceived(this, args);
            }
        }
    }
}

File: Client/Java/gamify-java-client/src/main/java/ar/com/mdf/gamify/managers/GameMoveManager.java
package ar.com.mdf.gamify.managers;

import java.net.URISyntaxException;

import ar.com.mdf.gamify.client.IGamifyClient;
import ar.com.mdf.gamify.enums.NotificationType;
import ar.com.mdf.gamify.enums.RequestType;
import ar.com.mdf.gamify.exceptions.NotConnectedException;
import ar.com.mdf.gamify.factory.IClientFactory;
import ar.com.mdf.gamify.objects.notification.INotificationObject;
import ar.com.mdf.gamify.objects.request.IRequestObject;
import ar.com.mdf.gamify.observers.IGameObserver;
import ar.com.mdf.gamify.services.GameListener;
import ar.com.mdf.gamify.services.GameService;
import ar.com.mdf.gamify.services.IGameListener;
import ar.com.mdf.gamify.services.IGameService;

public class GameMoveManager<TGameMoveRequest extends IRequestObject, TMoveResultNotification extends INotificationObject, TMoveNotification extends INotificationObject> {
	private IGameService<TGameMoveRequest, TMoveResultNotification> gameMoveService;
    private IGameListener<TMoveNotification> gameMoveListener;
    private IGamifyClient client;
    
    public GameMoveManager(IClientFactory clientFactory, Class<TMoveResultNotification> resultNotificationClass, Class<TMoveNotification> moveNotificationClass) throws NotConnectedException {
		try {
			this.client = clientFactory.getClient();
		} catch (URISyntaxException e) {
			throw new NotConnectedException(e);
		}
		this.gameMoveService = new GameService<TGameMoveRequest, TMoveResultNotification>(RequestType.GameMove, NotificationType.GameMoveResult, this.client, resultNotificationClass);
		this.gameMoveListener = new GameListener<TMoveNotification>(NotificationType.GameMove, moveNotificationClass);
	}
    
    public void suscribeMoveResult(IGameObserver<TMoveResultNotification> observer){
		this.gameMoveService.suscribe(observer);
	}
    public void suscribeMove(IGameObserver<TMoveNotification> observer){
		this.gameMoveListener.suscribe(observer);
	}
    
	public void sendMove(TGameMoveRequest moveRequest) {
		this.gameMoveService.send(moveRequest);
	}
}
